geos: dlopen libgeos.so before libgeos_c.so

otan opened the pull request:
We need to dlopen libgeos.so before libgeos_c.so when copying the
libraries directly to be copied into a tarball.

Also change the default location for these .so files to be
`/usr/local/lib/cockroach`, which is in line with other example
libraries seen (e.g. `/usr/local/lib/postgres`).

This PR is in preparation for a Docker update and making GEOS work with
roachprod utilities.

Release note (general change): The docker image that is
available with CockroachDB will also include the GEOS files necessary to
work with Geospatial types in CockroachDB.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/50362)
<!-- Reviewable:end -->

petermattis reviewed (commented):
Nit: commit message: `s/preperation/preparation/g`

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/50362#-:-MA35-ZI37cBJTRoayYU:b-ipz1a5)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @otan and @petermattis)

___
*[pkg/geo/geos/geos.go, line 78 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/50362#-MA35A8TE1dJwooTCOIr:-MA35A8U2g4vGWjjv1mJ:bc1u5vi) ([raw file](https://github.com/cockroachdb/cockroach/blob/8c9515373f71f68c1d0da6f7db2ac57fd1575f7a/pkg/geo/geos/geos.go#L78)):*
> ```Go
> 
> // LibraryLocation contains the locations of the relevant GEOS library files.
> type LibraryLocation struct {
> ```

Looks like you require `Dir(GEOSCFile) == Dir(GEOSFile)`. Would it be useful to simplify this down to a single directory? Feel free to ignore this suggestion as I only have time to skim this PR at the moment.
___
*[pkg/geo/geos/geos.go, line 140 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/50362#-MA34MF00WxKE51YZ3HN:-MA34MF00WxKE51YZ3HO:b1j841k) ([raw file](https://github.com/cockroachdb/cockroach/blob/8c9515373f71f68c1d0da6f7db2ac57fd1575f7a/pkg/geo/geos/geos.go#L140)):*
> ```Go
> 		if _, err := os.Stat(GEOSCFilePath); err == nil {
> 			geosPath := filepath.Join(cwd, "lib", "libgeos."+ext)
> 			if _, err := os.Stat(geosPath); err == nil {
> ```

Probably worth logging an error if we can find `libgeos_c.*` in a directory, but not `libgeos.*`. I imagine that might be a configuration error.


<!-- Sent from Reviewable.io -->

otan reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/50362)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @petermattis)

___
*[pkg/geo/geos/geos.go, line 78 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/50362#-MA35A8TE1dJwooTCOIr:-MA39Bcj6kabDlI_X4dX:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/8c9515373f71f68c1d0da6f7db2ac57fd1575f7a/pkg/geo/geos/geos.go#L78)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Looks like you require `Dir(GEOSCFile) == Dir(GEOSFile)`. Would it be useful to simplify this down to a single directory? Feel free to ignore this suggestion as I only have time to skim this PR at the moment.
</blockquote></details>

Done.
___
*[pkg/geo/geos/geos.go, line 140 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/50362#-MA34MF00WxKE51YZ3HN:-MA3CqGF1NdzhOx28m63:b-hylvqe) ([raw file](https://github.com/cockroachdb/cockroach/blob/8c9515373f71f68c1d0da6f7db2ac57fd1575f7a/pkg/geo/geos/geos.go#L140)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Probably worth logging an error if we can find `libgeos_c.*` in a directory, but not `libgeos.*`. I imagine that might be a configuration error.
</blockquote></details>

This is a best effort scan from the CI standpoint where we do not move the geos files to `/usr/local/lib/cockroach` I think it's okay if one is missing but not the other in this case (we should always check the flag provided location anyway).


<!-- Sent from Reviewable.io -->

otan committed:
geos: dlopen libgeos.so before libgeos_c.so

We need to dlopen libgeos.so before libgeos_c.so when copying the
libraries directly to be copied into a tarball.

Also change the default location for these .so files to be
`/usr/local/lib/cockroach`, which is in line with other example
libraries seen (e.g. `/usr/local/lib/postgres`).

This PR is in preparation for a Docker update and making GEOS work with
roachprod utilities.

Release note (general change): The docker image that is
available with CockroachDB will also include the GEOS files necessary to
work with Geospatial types in CockroachDB.

petermattis reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/50362#-:-MA6MG8sEeWu4B3Ff00V:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->

otan commented:
bors r=petermattis

TFTR!
